object-store-rs/src/buffered.rs

```diff
@@ -0,0 +1,212 @@
+use std::io::SeekFrom;
+use std::sync::Arc;
+
+use object_store::buffered::BufReader;
+use pyo3::exceptions::PyStopAsyncIteration;
+use pyo3::prelude::*;
+use pyo3_object_store::error::{PyObjectStoreError, PyObjectStoreResult};
+use pyo3_object_store::PyObjectStore;
+use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncSeekExt, Lines};
+use tokio::sync::Mutex;
+
+use crate::get::PyBytesWrapper;
+use crate::runtime::get_runtime;
+
+#[pyfunction]
+pub(crate) fn open(
+    py: Python,
+    store: PyObjectStore,
+    path: String,
+) -> PyObjectStoreResult<PyReadableFile> {
+    let store = store.into_inner();
+    let runtime = get_runtime(py)?;
+    let meta = py.allow_threads(|| runtime.block_on(store.head(&path.into())))?;
+    Ok(PyReadableFile(Arc::new(Mutex::new(BufReader::new(
+        store, &meta,
+    )))))
+}
+
+#[pyfunction]
+pub(crate) fn open_async(py: Python, store: PyObjectStore, path: String) -> PyResult<Bound<PyAny>> {
```

Comment:
I'm not sure you want to bother with open_async for now - if it's to be a file-like API, may as well just do that, and only give the blocking methods. It is a stateful object, after all.

Reply:
Presently, this PR only provides async methods, not blocking methods. The API is meant to be similar to https://github.com/Tinche/aiofiles.

It would be a good enhancement to have two separate Python classes, one returned by `open` and another returned by `open_async`. Then the methods on the object returned by `open_async` would be named `read()` (without the `async` suffix) but still return coroutines.

Reply:
OK, so not really like https://github.com/fsspec/filesystem_spec/blob/master/fsspec/asyn.py#L911 (implementations elsewhere)

Reply:
No, sorry, this wasn't modeled _after_ fsspec, but it could be used in the fsspec wrapper.